Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -721,13 +709,54 @@ namespace Babylon
             }};
 
         auto vertexShader = bgfx::createShader(bgfx::copy(shaderInfo.VertexBytes.data(), static_cast<uint32_t>(shaderInfo.VertexBytes.size())));
-        InitUniformInfos(vertexShader, shaderInfo.UniformStages, program->UniformInfos, program->UniformNameToIndex);
-        program->VertexAttributeLocations = std::move(shaderInfo.VertexAttributeLocations);
+        InitUniformInfos(vertexShader, shaderInfo.UniformStages, program.UniformInfos, program.UniformNameToIndex);
+        program.VertexAttributeLocations = std::move(shaderInfo.VertexAttributeLocations);
 
         auto fragmentShader = bgfx::createShader(bgfx::copy(shaderInfo.FragmentBytes.data(), static_cast<uint32_t>(shaderInfo.FragmentBytes.size())));
-        InitUniformInfos(fragmentShader, shaderInfo.UniformStages, program->UniformInfos, program->UniformNameToIndex);
+        InitUniformInfos(fragmentShader, shaderInfo.UniformStages, program.UniformInfos, program.UniformNameToIndex);
+
+        return bgfx::createProgram(vertexShader, fragmentShader, true);
+    }
 
-        program->Handle = bgfx::createProgram(vertexShader, fragmentShader, true);
+    Napi::Value NativeEngine::CreateProgram(const Napi::CallbackInfo& info)
+    {
+        const std::string vertexSource = info[0].As<Napi::String>().Utf8Value();
+        const std::string fragmentSource = info[1].As<Napi::String>().Utf8Value();
+        ProgramData* program = new ProgramData{};
```

Comment:
This will leak memory if CreateProgramInternal throws. If we make it such that CreateProgramInternal doesn't know about ProgramData, then we can create the ProgramData right before the Napi::Pointer which should fix it.

Reply:
Creating the `ProgramData` later means `CreateProgramInternal` wouldn't be able to access it. Instead, I've set it to delete `program` before throwing in `CreateProgram`. Hopefully in `CreateProgramAsync` no delete is needed because it does actually return even on a compile error. Please let me know if this is wrong.

Reply:
I pushed the changes we discussed (creating the data inside `CreateProgramInternal` and putting it into a `unique_ptr`). I gather that the `unique_ptr` will be destructed if `CreateProgramInternal` hits an exception before returning, so hopefully there's no more memory leak with this.